ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/JUnitLaunchRequirements.java

```diff
@@ -0,0 +1,164 @@
+		addAbsentRequirements(requiredPlugins, addedRequirements, allBundles, allModels);
+
+		Set<BundleDescription> requirementsOfRequirements = DependencyManager.findRequirementsClosure(addedRequirements);
+		addAbsentRequirements(requirementsOfRequirements, allBundles, allModels);
+	}
+
+	private static void addAbsentRequirements(Collection<String> requirements, Set<BundleDescription> addedRequirements, Map<String, List<IPluginModelBase>> allBundles, Map<IPluginModelBase, String> allModels) throws CoreException {
+		for (String id : requirements) {
+			List<IPluginModelBase> models = allBundles.computeIfAbsent(id, k -> new ArrayList<>());
+			if (models.stream().noneMatch(m -> m.getBundleDescription().isResolved())) {
+				IPluginModelBase model = JUnitLaunchRequirements.findRequiredPluginInTargetOrHost(id);
+				models.add(model);
+				BundleLauncherHelper.addDefaultStartingBundle(allModels, model);
+				if (addedRequirements != null) {
+					addedRequirements.add(model.getBundleDescription());
+				}
+			}
+		}
+	}
+
+	private static void addAbsentRequirements(Set<BundleDescription> toAdd, Map<String, List<IPluginModelBase>> allBundles, Map<IPluginModelBase, String> allModels) throws CoreException {
+		for (BundleDescription requirement : toAdd) {
+			String id = requirement.getSymbolicName();
+			List<IPluginModelBase> models = allBundles.computeIfAbsent(id, k -> new ArrayList<>());
+			boolean replace = !models.isEmpty() && models.stream().noneMatch(m -> m.getBundleDescription().getVersion().equals(requirement.getVersion()));
+			if (replace || models.stream().noneMatch(m -> m.getBundleDescription().isResolved())) {
+				IPluginModelBase model = JUnitLaunchRequirements.findRequiredPluginInTargetOrHost(requirement);
+				if (replace) {
+					String startLevel = null;
+					for (IPluginModelBase m : models) {
+						startLevel = allModels.remove(m);
+					}
+					models.clear();
+					allModels.put(model, startLevel);
+				}
+				models.add(model);
+				BundleLauncherHelper.addDefaultStartingBundle(allModels, model);
+			}
+		}
+	}
+
+	public static Set<BundleDescription> junit5PlatformRequirements() throws CoreException {
+		// add launcher and jupiter.engine to support @RunWith(JUnitPlatform.class)
+		String[] requiredPlugins = {"junit-platform-launcher", //$NON-NLS-1$
+				"junit-jupiter-engine", //$NON-NLS-1$
+		};
+		Set<BundleDescription> descriptions = new LinkedHashSet<>();
+		for (String id : requiredPlugins) {
+			IPluginModelBase model = findRequiredPluginInTargetOrHost(id, JUNIT5_VERSIONS);
+			if (model != null) {
+				BundleDescription description = model.getBundleDescription();
+				descriptions.add(description);
+			}
+		}
+		return descriptions;
+	}
+
+	public static IPluginModelBase findRequiredPluginInTargetOrHost(String id) throws CoreException {
+		IPluginModelBase model = findRequiredPluginInTargetOrHost(id, () -> PluginRegistry.findModel(id));
+		return model;
```

Comment:
We can save the line here and below and immediately return the result:
```suggestion
		return findRequiredPluginInTargetOrHost(id, () -> PluginRegistry.findModel(id));
```